### DISCUSSION TITLE
The position where you write *optimizer.zero_grad()*

### DISCUSSION THREAD

**Discussion by Reetos** [mrdbourke/pytorch-deep-learning#363] [General]
In the code below, when is the optimizer called and calculate the gradients?
From what I have understood, optimizer.zero_grad() is called to turn the gradients zero, but I do not know when the gradients were calculated before zero_grad() function was called in this code.

Or, in this case, was zero_grad() called just to instantiate the gradients with the value of zero, rather than to reset the gradients that have been calculated?

```
for epoch in range(epochs):
    ### Training
    model_1.train() # train mode is on by default after construction

    # 1. Forward pass
    y_pred = model_1(X_train)

    # 2. Calculate loss
    loss = loss_fn(y_pred, y_train)

    # 3. Zero grad optimizer
    optimizer.zero_grad()

    # 4. Loss backward
    loss.backward()

    # 5. Step the optimizer
    optimizer.step()
```
